### ISSUE TITLE
Automate static type analysis using mypy

### ISSUE THREAD

**Issue by data-sync-user** [mozilla-services/merino-py#42]
Incorporate mypy into merino-py automated checks in order to verify type hinting.

**Note:** 

* The mypy dependency and configuration are already present in the `pyproject.toml`
* There are currently errors when running the mypy

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/CONSVC-1955)
┆epic: Merino-py enhancements

**Comment by data-sync-user** [comment]
➤ Raphael Pierzina commented:

The mypy type checker is detecting *102 errors in 20 files* in merino-py for me locally on the latest main branch with the checked-in configuration. I strongly recommend we complete this ticket in the next sprint.

For additional context please see my message in Slack.

**Comment by data-sync-user** [comment]
➤ Taddes Korris commented:

Thanks Raphael Pierzina . I will start working on this today to resolve those issues and integrate the process into CI checks and the pre-commit hook. Appreciate you pointing this out and raising its priority.  I will update as soon as I have some significant progress complete.

**Comment by data-sync-user** [comment]
➤ Taddes Korris commented:

Hi Nan Jiang , Raphael Pierzina , Wil Stuckey ! As we discussed in our meetings, I wanted to put some docs on this ticket for us to at least keep some track of the Mypy revisions. I am happy to do as much of it as I can, but we did come to the realization that if someone has to review some 80 - 100 changes like this, things could get dicey and we introduce a bug. However, I do see that many issues were resolved with several PRs between us all. Plus, each person has some specific context related to the different libraries and middleware that have been brought in, so if a refactor is needed, it’s best it be handled by that person who initially implemented that functionality.  So, we can break this into sub-tasks by keeping track here, either broken up based on the directory structure, specific modules, or whatever you think is suitable.  I will post a separate comment shortly with some oddities, namely that Mypy is still very buggy when parsing return statements in a match statement, to the point where it doesn’t even detect one (I have seen the open issue in their Github), as well as some stub issues with the geoip2 module and not making sense of compound return types.

Either way, if we just keep track of things it’ll be a good sanity check for us all. Most of all, I don’t think we want deployment to be determined by the results of Mypy, as Philip Jenvey made some excellent points about diminishing returns due to the challenges Mypy introduces for annoying edge-cases. Any additional thoughts or ideas are welcome here!
